xray/core/dataset.py

```diff
@@ -1805,5 +1805,88 @@ def apply_over_both(lhs_data_vars, rhs_data_vars, lhs_vars, rhs_vars):
         ds._variables.update(new_vars)
         return ds
 
+    def diff(self, dim, n=1, coord_new='upper'):
+        """Calculate the n-th order discrete difference along given axis.
+
+        Parameters
+        ----------
+        dim : str, optional
+            Dimension over which to calculate the finite difference.
+
+        n : int, optional
+            The number of times values are differenced.
+
+        coord_new : str, optional
+            The new coordinate in dimension ``dim`` will have the
+            values of either the minuend's or subtrahend's coordinate
+            for values 'upper' and 'lower', respectively.  Other
+            values are not supported.
+
+        Returns
+        -------
+        difference : same type as caller
+            The n-th order finite differnce of this object.
+
+        Examples
+        --------
+        >>> ds = xray.Dataset({'foo': ('x', [5, 5, 6, 6])})
+        >>> ds.diff('x')
+        <xray.Dataset>
+        Dimensions:  (x: 3)
+        Coordinates:
+          * x        (x) int64 1 2 3
+        Data variables:
+            foo      (x) int64 0 1 0
+        >>> ds.diff('x', 2)
+        <xray.Dataset>
+        Dimensions:  (x: 2)
+        Coordinates:
+        * x        (x) int64 2 3
+        Data variables:
+        foo      (x) int64 1 -1
+
+        """
+        if n == 0:
+            return self
+        if n < 0:
+            raise ValueError('order `n` must be non-negative but got {}'
+                             ''.format(n))
+
+        # prepare slices
+        kwargs_start = {dim: slice(None, -1)}
+        kwargs_end = {dim: slice(1, None)}
+        start = self.isel(**kwargs_start)
+        end = self.isel(**kwargs_end)
+
+        # prepare new coordinate
+        if coord_new == 'upper':
+            start.coords[dim] = end.coords[dim]
+        elif coord_new == 'lower':
+            end.coords[dim] = start.coords[dim]
+        else:
+            raise ValueError('The \'coord_new\' argument has to be either '
+                             '\'upper\' or \'lower\'')
+
+        variables = OrderedDict()
+        for name, var in iteritems(self._variables):
+            if name not in variables.keys():
+                variables[name] = var
+            if dim in var.dims or not var.dims:
+                if name not in self.coords:
+                    variables[name] = end[name] - start[name]
+                    variables[dim] = variables[name].coords[dim]
```

Comment:
Let me test your PR and debug this. My guess is that some of the variables were ending up with inconsistent dimension sizes (the Dataset constructor resolves that by doing an outer join of the index labels).